game_mechanics.py

```diff
@@ -133,27 +231,36 @@ def flip_tiles(board: np.ndarray, move: Tuple[int, int], current_player: int) ->
     return board
 
 
+def get_possible_moves(board: np.ndarray) -> set:
+
+    board_dim = len(board)
+    zero_idx = board == 0
+    # Few empty, iterate through these (1.5 threshold chosen empirically for speed)
+    if np.sum(zero_idx) < board_dim**2 / 1.5:
+        idx_check = np.where(zero_idx)
+        to_check = set(zip(idx_check[0], idx_check[1]))
+    else:
+        to_check = set()
+        idx_pieces = np.where(~zero_idx)
+        for pos in zip(idx_pieces[0], idx_pieces[1]):
+            idxs = idx_surrounding(board, pos)
+            to_check.update(list(zip(idxs[0], idxs[1])))
+
+    return to_check
+
+
 def has_legal_move(board: np.ndarray, current_player: int) -> bool:
     """Checks whether current_player has any legal move to make."""
-    board_dim = len(board)
-    for row in range(board_dim):
-        for col in range(board_dim):
-            move = (row, col)
-            if is_legal_move(board, move, current_player):
-                return True
+    possible_moves = get_possible_moves(board)
+    for move in possible_moves:
+        if is_legal_move(board, move, current_player):
+            return True
     return False
```

Comment:
:bomb: :boom: :boom: :boom: :boom: :boom: :boom: :boom: 
```suggestion
    return any(is_legal_move(board, move, current_player) for move in possible_moves)
```